### DISCUSSION TITLE
Proposed changes to examples/

### DISCUSSION THREAD

**Discussion by jones-michael-s** [automaticanalysis/automaticanalysis#327] [General]
I'm getting ready to post the tutorial to the website and it feels like the current organization of the examples/ directory could also use some attention.

Current organization: 

`cbu/ demos/ tutorials/ utilities/`

I think the following changes would improve things:

1) changes to tutorials/

- move aa_example_helloworld_model.m to demos/

aa_example_helloworld_model.m was something I was going to use for the tutorial based on the ds000114 motor task. In the end I went with the SPM auditory task (mainly because it's simpler and the SPM manual provides additional documentation). Also, "helloworld" is snarky programmer humor that maybe isn't helpful. So I'd like to offer this as demos/ds000114_motor.m/xml. It complements bids_ds000114_aa.m/xml currently in demos which showcases all the tasks in the dataset as well as some other aa features, whereas my example is rather more barebones.

- delete tutorial_1_aa_setup.m/xml

This is mainly parameter file setup and while I think is the setup tool is a great addition, once the tutorial is posted we can just tell the user to run the parameter file setup tool before running the tutorial script. Putting parameter setup in the userscript feels weird -- that's not an "analysis" task; it's a "do once and forget about it" task. Therefore I would delete this file.

Additionally, I'd like to add a wrapper (create_basic_parameterfile.m) in utilities/ that just passes the defaults to aas_create_parameter_xml so the user doesn't have to know about these details.

- rename tutorial_2_SPM_CH30.m/xml

I would go with either "tutorial.m/xml" or maybe keep the name "SPM_CH30.m/xml".

Also, IIRC this file is currently organized as a function so that it can be included in the CI. This is a good idea (we always want to make sure the tutorial code is working because its the first thing a new user will see). However, like parameter set-up, it feels weird to me. I've never structured a userscript as a function (except for testing) and while its a small thing, it's still one more thing that a new user doesn't need to know about. So I would propose adding a copy of SPM_CH30 to developer/testscripts (then tutorial_2_SPM_CH30.m/xml can simply be renamed tutorial.m/xml).

Also, I've kept automatic data download in the script but the user has to uncomment it. Autodownload is problematic at best (indeed, it crashed when I tried it). If aa crashes immediately the first time a new user tries it, that user isn't going to stick around. As such, I now explain in the tutorial text that autodownload can sometimes fail and include instructions for manual download, but they are free to try autodownload if they want.

2) Move utilities/ under aa_tools?

This isn't a huge deal, but having utilities/ under examples/ give the impression these are, well, example utilities -- that is, examples of utilities that a user may need to adapt for their own work. That's not what's in the directory -- they are aa utilities that happen to be used in the example scripts. So maybe move the directory to reflect this?

3) rename demos/? 

IIRC, the difference between demos/ and cbu/ is that all the scripts in demos/ use open data. So maybe rename the directory to "open_data/" or "public_data/"? Or simpler: just move all of the stuff in demos/ up one level (i.e., to examples/). 

Not a big deal, it's just confusing to have "example" scripts and "demo" scripts. How is a demo different than an example?

4) Finally, there's a bunch of different naming conventions used in the example scripts. Some use an aa_ prefix. Some use a demo_ prefix. Some use "tasklist" or "user[script]" as part of the name. Some use "aa" and "tasklist". 

Maybe more consistent naming would help a new aa user? (unless this would break other stuff like CI)

For example, instead of aa_example_denoising.m/xml why not just denoising.m/xml? We already know its an aa script (because its in the aa repo) and we know its an example (because its in the examples/ directory). We know denoising.m is a userscript (because .m). We know denoising.xml is a tasklist (because .xml).

**Comment by tiborauer** [comment]
#### Ad 1.: 
##### Re aa_example_helloworld_model:
The `demos` folder was meant for demonstrating more advanced use cases beyond the tutorials. If you say `aa_example_helloworld_model` is more simple, then it may be redundant and could be deleted.

##### Re tutorial_1_aa_setup and create_basic_parameterfile:
+1

##### Re tutorial_2_SPM_CH30:
I would keep the SPM_CH30 as a reference to the use case.

I am OK having two copies, (script and function). I wonder whether the function could simply be a wrapper calling the script to avoid redundancy and make the link between the two more explicit.

I am OK with (un)commenting autodownload. You may even add a configuration variable to the top of the script so user rather enable/disable autodownload than comment/uncomment code (which I find less elegant). 

#### Ad 2.
+2 (Very good idea!)

#### Ad 3.
As in Ad 1., we decided for a clear distinction bewteen `tutorials` and `demos` - the former for introductory use cases, the latter for more advanced use cases. Relying on open data has also been set as a goal for all examples; therefore I would not use it as a distinctive feature. I consider the `cbu` examples as a WIP and they should all adapted to an open data at some point. If you feel keeping them in the example somewhat confusing (I certainly do), then we may remove them completely and re-implement them if we know which data to use for them.

#### Ad 4.
+2
I think if they are scripts then they are not part of the CI, but we will know if it breaks any CI. :)

**Comment by jones-michael-s** [comment]
Ok, I'll put together a PR with these changes.

Thanks for the feedback.
